src/main/java/com/amazonaws/encryptionsdk/kms/KmsMasterKeyProviderBuilder.java

```diff
@@ -0,0 +1,125 @@
+ * </pre>
+ *
+ * Example, specifying a region:
+ * <pre>
+ *     KmsMasterKeyProvider keyProvider = KmsMasterKeyProviderBuilder.standard()
+ *       .withRegion("us-east-1")
+ *       .build();
+ * </pre>
+ */
+public class KmsMasterKeyProviderBuilder extends AwsClientBuilder<KmsMasterKeyProviderBuilder, KmsMasterKeyProvider> {
+
+    private List<String> keyIds;
+
+    /**
+     * @return a new {@link KmsMasterKeyProviderBuilder} with the defaults set.
+     */
+    public static KmsMasterKeyProviderBuilder standard() {
+        return new KmsMasterKeyProviderBuilder();
+    }
+
+    /**
+     * @return a KmsMasterKeyProvider using the {@link AWSKMSClientBuilder} defaults, and an empty key ID list.
+     */
+    public static KmsMasterKeyProvider defaultProvider() {
+        return standard().build();
+    }
+
+    private KmsMasterKeyProviderBuilder() {
+        super(new ClientConfigurationFactory());
+    }
+
+    /**
+     * Sets the region to be used by the client. Overrides any previously set region.
+     */
+    public KmsMasterKeyProviderBuilder withRegion(Region region) {
+        return withRegion(region.getName());
+    }
+
+    /**
+     * Sets the region using a keyId. This will override any previously set regions.
+     */
+    public KmsMasterKeyProviderBuilder withKeyId(String keyId) {
+        withKeyIds(Collections.singletonList(keyId));
+
+        withRegion(KmsMasterKeyProvider.getStartingRegion(keyId));
+
+        return this;
+    }
+
+    /**
+     * Adds {@code keyIds}, but does not use them to set the region.
+     */
+    public KmsMasterKeyProviderBuilder withKeyIds(List<String> keyIds) {
+        if (this.keyIds == null) {
+            this.keyIds = new ArrayList<>();
+        }
+
+        this.keyIds.addAll(keyIds);
+
+        return this;
+    }
+
+    /**
+     * Builds a {@link KmsMasterKeyProvider} using the information it was built with, or with defaults where necessary.
+     */
+    @Override
+    public KmsMasterKeyProvider build() {
+        return build(getSyncClientParams());
+    }
+
+    private KmsMasterKeyProvider build(AwsSyncClientParams clientParams) {
+        keyIds = (keyIds == null) ? Collections.emptyList() : keyIds;
+
+        AWSKMSClient client = new AWSKMSClient(clientParams.getCredentialsProvider(),
+                clientParams.getClientConfiguration(),
+                clientParams.getRequestMetricCollector());
+
+        Region region = determineRegion();
+
+        return new KmsMasterKeyProvider(client,
+                region,
+                keyIds);
+    }
+
+    private Region determineRegion() {
+        Region region = RegionUtils.getRegion(this.getRegion());
+
+        final AwsRegionProvider regionProvider = new DefaultAwsRegionProviderChain();
+
+        if (region != null) {
+            return region;
+        } else {
+            final String regionName = regionProvider.getRegion();
+            if (regionName != null) {
```

Comment:
instead of using US_WEST_2 as default, could we use Regions.getCurrentRegion()?

Reply:
This looks like it's only a fallback in the case where the `DefaultAwsRegionProviderChain` fails. Which is reasonable enough,

Reply:
This looks like a fallback in the case where the SDK's default region selection logic fails, which is reasonable enough.

Reply:
@SalusaSecondus,
in Regions.java inside aws-java-sdk-core:
```java
 /**
     * The default region that new customers in the US are encouraged to use
     * when using AWS services for the first time.
     */
    public static final Regions DEFAULT_REGION = US_WEST_2;
```

@bdonlan agree with you. looks good to me.